### PR TITLE
New DeviceManager config object

### PR DESCRIPTION
**Pull request by gcatron**
*Description*: Introduced new DeviceConfig struct to hold device configuration information for device managers. Changed factory and constructors to accept new struct instead of the current name parameter.
*Testing*: ninja test and resnet-runtime
*Documentation*: NA

### CODE REVIEW

File: include/glow/Runtime/RuntimeTypes.h [hunk 1]
```diff
@@ -81,6 +81,15 @@ struct DAGNodeList {
   std::vector<std::unique_ptr<DAGNode>> nodes;
 };
 
+/// This is the base struct for DeviceManager configurations. The base struct
+/// contains a string for the device name. Any specific device can extend this
+/// struct to contain additional information to identify and configure the
+/// device manager.
+struct DeviceConfig {
+  /// A string containing the name of the device the device manager controls.
+  std::string deviceName;
```
Comment (nickgg) [inline include/glow/Runtime/RuntimeTypes.h]: You have deviceName as a parameter here but you've deleted every point in the code that set a device name as far as I can tell. Do we need this?

Comment (gcatron) [reply include/glow/Runtime/RuntimeTypes.h]: Probably not, I'll go ahead and remove it.

File: lib/Backends/Interpreter/InterpreterDeviceManager.h [hunk 2]
```diff
@@ -40,9 +40,9 @@ class InterpreterDeviceManager : public QueueBackedDeviceManager {
   const uint64_t functionCost_{1};
 
 public:
-  InterpreterDeviceManager(llvm::StringRef name = "unnamed",
+  InterpreterDeviceManager(std::unique_ptr<DeviceConfig> config = nullptr,
```
Comment (SplitInfinity) [inline lib/Backends/Interpreter/InterpreterDeviceManager.h]: Since we know the type of backend at this point, I wonder if all of these constructors should take in subclass pointers instead. My goal is to help quickly catch mistakes like this that might be made while writing an example or unit test:

```
auto dm = std::make_unique<InterpreterDeviceManager>(cpuDeviceConfigPtr);
```

File: lib/Backends/DeviceManagers.cpp [hunk 3]
```diff
@@ -29,46 +29,52 @@ namespace runtime {
 /// when you define a new DeviceManager.
 
 /// Create a new instance of the interpreter Device.
-DeviceManager *createInterpreterDeviceManager(llvm::StringRef name);
+DeviceManager *
+createInterpreterDeviceManager(std::unique_ptr<DeviceConfig> config = nullptr);
```
Comment (SplitInfinity) [inline lib/Backends/DeviceManagers.cpp]: Why did you remove the name field? For some vendors, that might be used to differentiate between different devices (e.g. by PCIe bus name).

Comment (gcatron) [reply lib/Backends/DeviceManagers.cpp]: That would go into the derived config object for that vendor. Since it's not global to all types I removed it.

File: examples/tracing-compare.cpp [hunk 4]
```diff
@@ -106,8 +106,7 @@ int main(int argc, char **argv) {
 
   std::array<DeviceManager *, supportedBackends.size()> devices;
   for (unsigned i = 0, e = supportedBackends.size(); i < e; ++i) {
-    devices[i] = DeviceManager::createDeviceManager(supportedBackends[i],
-                                                    "tracing-compare");
+    devices[i] = DeviceManager::createDeviceManager(supportedBackends[i]);
     devices[i]->init();
```
[the remark this answers is not in the document]
Comment (gcatron) [reply examples/tracing-compare.cpp]: created issue #2485

File: lib/Backends/Interpreter/InterpreterDeviceManager.h (hunk 2 above)
Comment (gcatron) [reply lib/Backends/Interpreter/InterpreterDeviceManager.h]: That still requires a check at runtime. (Ideally we'd want to dyn_cast and check). This can also be handled by the device manager that is using the config. Pushing the check down to the manager means we only need a config object defined for a manager that needs one.

Comment (nickgg) [reply lib/Backends/Interpreter/InterpreterDeviceManager.h]: I think we will need a runtime check, but as discussed offline we don't have RTTI available.

Do you think it's worth adding a check for devices like Interpreter/CPU which probably wont use anything in the config?

Comment (gcatron) [reply lib/Backends/Interpreter/InterpreterDeviceManager.h]: I don't think so, if we check in the device managers that actually use the config we only have to check when it is actually used. I'll add the mechanism to check kind to this PR.

**Comment by jackm321** [comment]
Looks good.
I also have some concerns like what were raised already here about passing a generic config type to each backend without any RTTI we just have to cast and hope we got the right type of config for the backend. That being said, I'm not sure what a better alternative would really look like.
